### PR TITLE
Enable softmax raising by default in IREE

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Also add a pass to canonicalize linalg ops to catch more cases of softmax patterns.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 1]
```diff
@@ -200,6 +201,7 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
       // - Remove unit-extent dimensions.
       .addPass(mlir::createConvertElementwiseToLinalgPass)
       .addPass(mlir::createLinalgFoldUnitExtentDimsPass)
+      .addPass(createRaiseSpecialOps)
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp:204]: Maybe better to move these also much earlier.... especially before const propagation.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks! mostly looks fine. Just a few comments.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 2]
```diff
@@ -756,6 +754,10 @@ void buildLLVMCPUCodegenPassPipeline(OpPassManager &passManager) {
   passManager.nest<ModuleOp>().addNestedPass<func::FuncOp>(
       createLLVMCPUMaterializeEncodingPass());
   passManager.addNestedPass<ModuleOp>(createBufferizeCopyOnlyDispatchesPass());
+  // Lower softmax early to have it go through the exisiting reduction lowering.
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:757]: Right, the reason why I put it there is to get the codegen from aggressive fusion out of the box. I can leave it to decompose but I'm guessing I'll get a perf regression without doing more. I'm trying to get it enabled without perf regression. Then ideally someone should pick it up to do something better on the cpu backend.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp:204]: Running it after `createLinalgFoldUnitExtentDimsPass` allows catching more cases as it put the ops in a more canonical form. Otherwise some front ends create unit dims while some others don't

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:757]: There shouldnt be a perf regression? Why do you think there would be?

[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:757]: because I'm guessing the reductions used to be optimized and now I'm replacing those by a softmax that won't be optimized. I'm guessing based on what is done on GPU backends but I could be wrong. I'm reverting that part if you think it won't be a problem.

[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: reverted it

[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: I had to move it back otherwise this causes stack explosion

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 3]
```diff
@@ -758,6 +756,11 @@ void buildLLVMCPUCodegenPassPipeline(OpPassManager &passManager) {
   passManager.nest<ModuleOp>().addNestedPass<func::FuncOp>(
       createLLVMCPUMaterializeEncodingPass());
   passManager.addNestedPass<ModuleOp>(createBufferizeCopyOnlyDispatchesPass());
+  passManager.nest<ModuleOp>().addNestedPass<func::FuncOp>(
+      IREE::LinalgExt::createDecomposeSoftmaxPass());
+  // Temporary solution to avoid large allocations due to softmax lowering.
+  passManager.nest<ModuleOp>().addNestedPass<func::FuncOp>(
+      createRematerializeParallelOpsPass());
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: Could we add some tests for this?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: what kind of test do you think we could add? Those passes are tested individually and there is an integration test for softmax already. The codegen pipeline tests would not include this part so I'm not sure what would be interested to test. What do you think?

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: I guess it should be fine then. I was thinking about a test to make sure that no other pass mess up whatever this pass should be decomposing.

**Review by dcaballe — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: You can try adding a pipeline test.

**Comment by hanhanW** [comment]
woohoo! nice improvement on x86 side!

**Comment by MaheshRavishankar** [comment]
> woohoo! nice improvement on x86 side!

Really? I missed it, which one?

**Comment by hanhanW** [comment]
> > woohoo! nice improvement on x86 side!
> 
> Really? I missed it, which one?

Looks like there are improvements on MobileNetV2 and V3 for single-threaded

image

**Comment by benvanik** [comment]
> Looks like there are improvements on MobileNetV2 and V3 for single-threaded

(would want those confirmed before we call mission accomplished - +/-10% swings are normal noise)

**Comment by MaheshRavishankar** [comment]
> > > woohoo! nice improvement on x86 side!
> > 
> > 
> > Really? I missed it, which one?
> 
> Looks like there are improvements on MobileNetV2 and V3 for single-threaded
> 
> image

What Ben said. That is probably noise.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: Correct me if I am wrong, but I think the decompose softmax should probably be running elementwise operation fusion as well. You can probably also use the cost function of the fusion to incorporate rematerialization with the `decomposeSoftmax` code. That probably gives you the lowering you need?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:763]: correct, I was planning to keep it for future clean up unless you think it makes things worse?

**Comment by hanhanW** [comment]
Noise mostly happens on Pixels. I seldom see noice on our x86 benchmark CI. but yeah, agree that we want these confirmed, maybe by running it for 10 times.

**Comment by dcaballe** [comment]
Yeah, x86 are usually more stable. I get no improvements/regressions most of the times if the PR doesn't impact x86. Of course, you run the benchmarks multiple times to be sure.

**Review by MaheshRavishankar — APPROVED** [review]
Looked at IR generated on CPU side for softmax. It looks good. So removing my blocker. Please prioritize the change to decompose code that will allow making the rematerialization in a more targeted manner by moving it into the decomposition itself.

**Comment by ThomasRaoux** [comment]
This mostly improves cases coming from pytorch at the time. There is a slightly different pattern we need for TF cases that should boost the benchmarks we have. I'm working on adding it shortly but it needs some matcher improvements to land first

**Comment by benvanik** [comment]
if y'all understand them and trust them that's the most important thing - that we have a 5.52% decrease on pixel 6 cpu listed as "improved latencies" and a 12.91% increase on pixel 6 cpu listed as "similar latencies" makes the numbers feel random - that it differs across targets compounds my issues with trying to build a mental model around the meaning of these things x_x
that's off topic, though, and sorry for the noise - if the numbers are real here then that's awesome 🥳
